### ISSUE TITLE
unexpected error messages in nc_test4/tst_vars2.c NC_ELATEFILL not being returned as expected!

### ISSUE THREAD

**Issue by edhartnett** [Unidata/netcdf-c#577]
Looking out the output of tst_vars2.c I see that there are some errors reported.

Additionally there is the serious question of why there were apparently errors, but the test still returns a successful exit code.

```
*** Testing netcdf-4 variable functions, even more.
*** Testing with classic format:
**** testing simple fill value attribute creation...Error at line 74: expecting NC_ELATEFILL but got No error
ok.
**** testing simple fill value with data read...ok.
**** testing fill value with one other attribute...ok.
**** testing fill value with three other attributes...ok.
**** testing fill value with simple example...ok.
**** testing fill value with example from cxx tests in v2 api...ok.
*** Testing with 64-bit offset format:
**** testing simple fill value attribute creation...Error at line 74: expecting NC_ELATEFILL but got No error
ok.
**** testing simple fill value with data read...ok.
**** testing fill value with one other attribute...ok.
**** testing fill value with three other attributes...ok.
**** testing fill value with simple example...ok.
**** testing fill value with example from cxx tests in v2 api...ok.
*** Testing with HDF5:
**** testing simple fill value attribute creation...ok.
**** testing simple fill value with data read...ok.
**** testing fill value with one other attribute...ok.
**** testing fill value with three other attributes...ok.
**** testing fill value with simple example...ok.
**** testing fill value with example from cxx tests in v2 api...ok.
*** Testing with HDF5, netCDF Classic Model:
**** testing simple fill value attribute creation...ok.
**** testing simple fill value with data read...ok.
**** testing fill value with one other attribute...ok.
**** testing fill value with three other attributes...ok.
**** testing fill value with simple example...ok.
**** testing fill value with example from cxx tests in v2 api...ok.
**** testing create order varids...ok.
**** testing simple variable renaming...ok.
**** testing dimension and variable renaming...ok.
**** testing endianness...ok.
**** testing chunking and the chunk cache...ok.
**** testing contiguous storage...ok.
**** testing extreme numbers dude...ok.
**** testing error codes for name clashes...ok.
**** testing error codes for name clashes some more...ok.
**** testing error codes for name clashes even more...ok.
**** testing error code for too-large chunks...ok.
**** testing that contiguous storage can't be turned on for vars with unlimited dims or filters...ok.
**** testing that contiguous storage can't be turned on for vars with unlimited dims or filters...ok.
*** Tests successful!
```

**Comment by edhartnett** [comment]
Well, this is a bit of a sticky wicket. Here's the code:

```
      printf("**** testing simple fill value attribute creation...");
      {
         int status;
         /* Create a netcdf-4 file with one scalar var. Add fill
          * value. */
         if (nc_create(FILE_NAME, cmode, &ncid)) ERR;
         if (nc_def_var(ncid, VAR_NAME, NC_BYTE, 0, NULL, &varid)) ERR;
         if (nc_put_att_schar(ncid, varid, _FillValue, NC_BYTE, 1, &fill_value)) ERR;
         if (nc_enddef(ncid)) ERR;
         if (nc_redef(ncid)) ERR;
         status = nc_put_att_schar(ncid, varid, _FillValue, NC_BYTE, 1, &fill_value);
         if (status != NC_ELATEFILL)
             printf("Error at line %d: expecting NC_ELATEFILL but got %s\n",__LINE__,nc_strerror(status));
         if (nc_close(ncid)) ERR;

```

There are two problems here:

1 - According to this code, the NC_ELATEFILL should be returned, but is not being returned.

2 - We have a test that depends on human scanning of the test output. Well that is not going to work! ;-)

I can easily change this to fix problem 2, but then the test will break. So I need to get a better understanding of what the NC_ELATEFILL error is all about - I think it is since my time on the project.

**Comment by edhartnett** [comment]
OK, looking at the code in libsrc/attr.m4, I see no attempt to catch this situation and return an NC_ELATEFILL.

I see that this is done in netCDF-4 (nc4attr.c), and tested in nc_test4/tst_vars2.c and other places.

**Comment by edhartnett** [comment]
I am not intimately familar with the classic code. @DennisHeimbigner is there an easy way to tell from this struct whether a variable has had any data written to it or not?

```
/*
 * NC variable: description and data
 */
typedef struct NC_var {
	size_t xsz;		/* xszof 1 element */
	size_t *shape; /* compiled info: dim->size of each dim */
	off_t *dsizes; /* compiled info: the right to left product of shape */
	/* below gets xdr'd */
	NC_string *name;
	/* next two: formerly NC_iarray *assoc */ /* user definition */
	size_t ndims;	/* assoc->count */
	int *dimids;	/* assoc->value */
	NC_attrarray attrs;
	nc_type type;		/* the discriminant */
	size_t len;		/* the total length originally allocated */
	off_t begin;
} NC_var;

```

**Comment by wkliao** [comment]
I believe this is also fixed in #731

**Comment by wkliao** [comment]
As for the error reporting, when expected status is not equal to NC_ELATEFILL, err should be increased by 1 and exit the program immediately (like other error handling). This part is not in #731.
